clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -307,34 +371,311 @@ impl FunctionType {
+            inner_type_check_type(
+                db,
+                contract_context,
+                atom_list_type.get_list_item_type(),
+                expected_list_type.get_list_item_type(),
+                clarity_version,
+                depth + 1,
+                tracker,
+            )?;
+        }
+        (
+            TypeSignature::TupleType(atom_tuple_type),
+            TypeSignature::TupleType(expected_tuple_type),
+        ) => {
+            for (name, atom_field_type) in atom_tuple_type.get_type_map() {
+                match expected_tuple_type.field_type(name) {
+                    Some(expected_field_type) => {
+                        inner_type_check_type(
+                            db,
+                            contract_context,
+                            atom_field_type,
+                            expected_field_type,
+                            clarity_version,
+                            depth + 1,
+                            tracker,
+                        )?;
+                    }
+                    None => {
+                        return Err(CheckErrors::TypeError(
+                            expected_type.clone(),
+                            atom_field_type.clone(),
+                        )
+                        .into())
                     }
                 }
             }
         }
-        Ok(returns.clone())
+        (
+            TypeSignature::CallableType(CallableSubtype::Trait(atom_trait_id)),
+            TypeSignature::CallableType(CallableSubtype::Trait(expected_trait_id)),
+        ) => {
+            if atom_trait_id != expected_trait_id {
+                let atom_trait = lookup_trait(
+                    db,
+                    contract_context,
+                    &atom_trait_id,
+                    clarity_version,
+                    tracker,
+                )?;
+                let expected_trait = lookup_trait(
+                    db,
+                    contract_context,
+                    expected_trait_id,
+                    clarity_version,
+                    tracker,
+                )?;
+                trait_check_trait_compliance(
+                    db,
+                    contract_context,
+                    &atom_trait_id,
+                    &atom_trait,
+                    expected_trait_id,
+                    &expected_trait,
+                    clarity_version,
+                    tracker,
+                )?;
+            }
+        }
+        (
+            TypeSignature::CallableType(CallableSubtype::Principal(contract_identifier)),
+            TypeSignature::CallableType(CallableSubtype::Trait(expected_trait_id)),
+        ) => {
+            let contract_to_check = db
+                .load_contract(&contract_identifier)
+                .ok_or(CheckErrors::NoSuchContract(contract_identifier.to_string()))?;
+            runtime_cost(ClarityCostFunction::AnalysisFetchContractEntry, tracker, 1)?;
```

Comment:
My only outstanding question for this PR is whether or not I need to use some size for this cost function other than 1. Should it be somehow based on the size of the `ContractAnalysis` structure, or should that size just be taken into account in the cost function? `AnalysisFetchContractEntry` is not used anywhere else, so I think we can just account for it in the cost function in costs-3, but this cost tracking is still a bit unclear to me.

Reply:
Per the blockchain meeting, the consensus was that this needed to be the size of the contract string (in bytes).

Reply:
Right, so I will pass the size of the contract string in the event one is found, else, I will pass a 1. Sound right?

Reply:
This has been resolved by passing the size of the public and read-only functions vectors instead, since the contract's source code (or its length) are not readily available.